### DISCUSSION TITLE
Constraint Specification

### DISCUSSION THREAD

**Discussion by corinwagen** [coltonbh/qcio#17] [Q&A] [answered]
Nice package! Pydantic is a solid choice for this. 

Is there currently a way to handle constraint specification for optimizations? The TeraChem manual (v1.9) suggests appending the geomeTRIC/DLFIND code to the end of the input file like so:

```
$constraint_set
bond 1.5 5_6 # Set bond length b/t atoms 5 and 6 to 1.5 Angstrom
angle 30 1_2_3 # Set 1-2-3 bond angle to 30 degrees
angle 45 2_3_4,5_6_7 # Set 2-3-4 and 5-6-7 bond angles to 45 degrees
$end
```

But I'm not sure how to pass this along -- is there a way to put this into ProgramInput.keywords?

**Comment by coltonbh** [comment, marked as answer]
Hi @corinwagen! Thanks for the kind words :)

The [qcop](https://github.com/coltonbh/qcop) package is where I've implemented support for various QC programs--like TeraChem or geomeTRIC--so you can find details there.

The constrained optimization IS supported if you use geomeTRIC as the main program and TeraChem as the subprogram running the energy/gradient calcs. You can see an example of how you could run this using `qcop` [here](https://github.com/coltonbh/qcop/blob/master/examples/geometric.py). Let me know if you have any other questions!

**Reply by corinwagen** [reply]
Ah, didn't understand this properly. Thanks!
